ext/drreg/drreg.c

```diff
@@ -2047,43 +2063,78 @@ drreg_event_restore_state_with_ilist(void *drcontext, bool restore_memory,
      * different slot) after app instr write, multiple reg restores due to app
      * instr read, aflags spilled to reg instead of slot (for x86 only), lazy
      * restores, multi-phase use of drreg leading to nested or overlapping
-     * spill regions, re-spills of native value by multiple phases, ...
-     * The strategy we follow here is to trace the app value of aflags and gpr
-     * as it is moved between spill slots and regs. At times the app value may
-     * be present in the gpr/aflags and one or more spill slots too, but later
-     * the gpr or any of the spill slots may be clobbered by a tool write or
-     * another spill respectively. At the end, we restore all live gprs/aflags
-     * that do not have their native app value.
+     * spill regions, re-spills of native value by multiple phases, spills by
+     * DR using non-drreg routines (to non-drreg spill slots or even stack).
+     *
+     * The strategy we follow here is to trace the app value of aflags and each
+     * gpr as it is moved between spill locations and regs. At times the app value
+     * may be present in the gpr/aflags and one or more spill locations too, but
+     * later the gpr or any of the spill locations may be clobbered by a tool write
+     * or another spill respectively. At the end, we restore the value of all non-
+     * native gprs/aflags for which we have some spill information. If we don't
+     * have any spill information for some non-native gpr/aflags, it probably
+     * means that the gpr/aflags were dead, therefore weren't spilled.
+     *
+     * Even though we do not need to restore gprs spilled by non-drreg spilling
+     * methods (like dr_save_reg or other methods like pushing to stack), we need
+     * to keep track of such spills/restores. This is needed to correctly track
+     * whether a gpr contains its app value currently or not, which is required to
+     * differentiate between app value and tool value spills, which is particularly
+     * important if drreg was used in multiple phases.
      */
+
+    /* Tracks which gpr's app value is stored by each spill slot. */
     reg_id_t spill_slot_to_reg[MAX_SPILLS];
-    /* Note that we do not need to track multiple spill regs for aflags. On x86
+    /* Tracks which gpr contains the aflags app value.
+     * Note that we do not need to track multiple spill regs for aflags. On x86
      * aflags can be spilled only to xax using lahf. On AArchXX, while they can
      * be read into any register using mrs, they are always spilled to a slot.
      */
     reg_id_t aflags_spill_reg;
+    /* Tracks which gpr contains the mcontext base. This is required to identify
+     * mcontext spills/restores.
```

Comment:
For 32-bit thread-private, there is no base: an absolute address is used.

The mcontext is used for regular spills, instead of TLS, in some cases: so "mcontext spills" might be confusing, since instr_is_reg_spill_or_restore will identify those in such cases.  (Never happens for 64-bit x86.)